### PR TITLE
feat: add `set_storage` [APE-1093]

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did

should be compatible with https://github.com/ApeWorX/ape/pull/1484, but not required to merge

### CODE REVIEW

File: ape_foundry/provider.py [hunk 1]
```diff
@@ -674,6 +674,18 @@ def set_code(self, address: AddressType, code: ContractCode) -> bool:
         self._make_request("anvil_setCode", [address, code])
         return True
 
+    def set_storage(self, address: AddressType, slot: int, value: HexBytes):
```
Comment (antazoey) [inline ape_foundry/provider.py:677]: i noticed it is using a different `HexBytes` here ... core is using `ethpm_types.HexBytes` but this is `hexbytes.HexBytes`.

All tests still pass for me when I make the swap though: https://github.com/ApeWorX/ape-foundry/pull/56

File: ape_foundry/provider.py [hunk 2]
```diff
@@ -674,6 +674,18 @@ def set_code(self, address: AddressType, code: ContractCode) -> bool:
...
+        def to_bytes32(val):
+            val_bytes = to_bytes(val)
+            if len(val_bytes) > 32:
+                raise ValueError(f"Value {val} must be <=32 bytes")
+            return HexBytes(b"\x00" * (32 - len(val_bytes))) + val_bytes
+
+        self._make_request(
```
Comment (antazoey) [inline ape_foundry/provider.py:684]: This should really have a test because of the `to_bytes32()` logic

**Review by antazoey — COMMENTED** [review]
probably should merge https://github.com/ApeWorX/ape-foundry/pull/56 first
and add a simple a test

File: ape_foundry/provider.py (hunk 2 above)
Comment (fubuloubu) [reply ape_foundry/provider.py:684]: Can put it in `utils` (or even `ape.utils`) it was really annoying to get this to work, I don't know why `eth-utils` doesn't provide this function

File: ape_foundry/provider.py [hunk 3]
```diff
@@ -674,6 +674,18 @@ def set_code(self, address: AddressType, code: ContractCode) -> bool:
...
+            "anvil_setStorageAt",
+            [address, to_bytes32(slot).hex(), to_bytes32(value).hex()],
```
Comment (banteg) [inline ape_foundry/provider.py:686]: i think hardhat requires slot to be encoded as quantity (hex with stripped leading zeros), need to check compatibility with different implementations.

**Review by banteg — APPROVED** [review]

File: ape_foundry/provider.py [hunk 4]
```diff
@@ -674,6 +674,18 @@ def set_code(self, address: AddressType, code: ContractCode) -> bool:
         self._make_request("anvil_setCode", [address, code])
         return True
 
+    def set_storage(self, address: AddressType, slot: int, value: HexBytes):
+        def to_bytes32(val):
+            val_bytes = to_bytes(val)
+            if len(val_bytes) > 32:
+                raise ValueError(f"Value {val} must be <=32 bytes")
+            return HexBytes(b"\x00" * (32 - len(val_bytes))) + val_bytes
```
Comment (banteg) [inline ape_foundry/provider.py:682]:
```suggestion
            return HexBytes(val_bytes.rjust(32, b"\x00"))
```

**Review by banteg — APPROVED** [review]

**Comment by fubuloubu** [comment]
Test failures are a regression of recent evm-trace/ape releases, not this PR

**Comment by fubuloubu** [comment]
New logic functions correctly, as noted the test failure is due to a regression already on `main`

File: ape_foundry/utils.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+from typing import Union
+
+from eth_utils import to_bytes, to_hex
+from ethpm_types import HexBytes
+
+
+# TODO: Upstream to ape core
```
Comment (fubuloubu) [inline ape_foundry/utils.py:7]: @antazoey can maybe do this later

**Comment by antazoey** [comment]
> New logic functions correctly, as noted the test failure is due to a regression already on `main`

hmm, there is a regression on main? I had thought that just the tests needed to be updated. I am looking at this now

**Review by antazoey — APPROVED** [review]

**Pull request merged**